core/src/main/java/org/sonarsource/sonarlint/core/container/connected/update/ModuleHierarchyDownloader.java

```diff
@@ -53,29 +55,37 @@ public ModuleHierarchyDownloader(SonarLintWsClient wsClient) {
    * @return Mapping of moduleKey -> relativePath from given module
    */
   public Map<String, String> fetchModuleHierarchy(String moduleKey) {
-    WsResponse response = wsClient.get("api/components/tree.protobuf?qualifiers=TRK,BRC&baseComponentKey=" + StringUtils.urlEncode(moduleKey));
-    try (InputStream stream = response.contentStream()) {
-      TreeWsResponse treeResponse = WsComponents.TreeWsResponse.parseFrom(stream);
+    List<Component> modules = new ArrayList<>();
 
-      // doesn't include root
-      Map<String, Component> components = treeResponse.getComponentsList().stream().collect(Collectors.toMap(Component::getId, Function.identity()));
+    int page = 0;
+    TreeWsResponse treeResponse;
+    do {
+      page++;
+      WsResponse response = wsClient.get("api/components/tree.protobuf?ps=500&p=" + page + "&qualifiers=BRC&baseComponentKey=" + StringUtils.urlEncode(moduleKey));
+      try (InputStream stream = response.contentStream()) {
+        treeResponse = WsComponents.TreeWsResponse.parseFrom(stream);
 
-      // component -> ancestorComponent. Doesn't include root
-      Map<Component, Component> ancestors = new HashMap<>();
-      for (Component c : treeResponse.getComponentsList()) {
-        ancestors.put(c, components.get(fetchAncestorId(c.getId())));
+        modules.addAll(treeResponse.getComponentsList());
+      } catch (IOException e) {
+        throw new IllegalStateException("Failed to load module hierarchy", e);
       }
+    } while (page * 500 < treeResponse.getPaging().getTotal());
```

Comment:
Magic number 500 is also embedded in the API query string, which might get overlooked when modifying this code in the future